netutil/netutil.go

```diff
@@ -54,3 +57,166 @@ func privateNetworkInterfaces(all []net.Interface, fallback []string, logger log
+	if !ipAddr.IsValid() {
+		return "", fmt.Errorf("no useable address found for interfaces %s", names)
+	}
+	if ipAddr.IsLinkLocalUnicast() {
+		level.Warn(logger).Log("msg", "using link-local address", "address", ipAddr.String())
+	}
+	return ipAddr.String(), nil
+}
+
+// getInterfaceAddresses is the standard approach to collecting []net.Addr from a network interface by name.
+func getInterfaceAddresses(name string) ([]netip.Addr, error) {
+	inf, err := net.InterfaceByName(name)
+	if err != nil {
+		return nil, err
+	}
+
+	addrs, err := inf.Addrs()
+	if err != nil {
+		return nil, err
+	}
+
+	// Using netip.Addr to allow for easier and consistent address parsing.
+	// Without this, the net.ParseCIDR() that we might like to use in a test does
+	// not have the same net.Addr implementation that we get from calling
+	// interface.Addrs() as above.  Here we normalize on netip.Addr.
+	netaddrs := make([]netip.Addr, len(addrs))
+	for i, a := range addrs {
+		prefix, err := netip.ParsePrefix(a.String())
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to parse netip.Prefix")
+		}
+		netaddrs[i] = prefix.Addr()
+	}
+
+	return netaddrs, nil
+}
+
+// filterBestIP returns an opinionated "best" address from a list of addresses.
+// A high quality address is one that is considered routable, and not in the link-local address space.
+// A low quality address is a link-local address.
+// When an IPv6 is enabled using enableInet6, an IPv6 will be preferred over an equivalent quality IPv4 address,
+// otherwise IPv6 addresses are guaranteed to not be returned from this function.
+// Loopback addresses are never selected.
+func filterBestIP(addrs []netip.Addr, enableInet6 bool) netip.Addr {
+	var invalid, inetAddr, inet6Addr netip.Addr
+
+	for _, addr := range addrs {
+		if addr.IsLoopback() {
+			continue
+		}
+
+		if addr.IsValid() {
+			if addr.Is4() {
+				// If we have already been set, can we improve on the quality?
+				if inetAddr.IsValid() {
+					if inetAddr.IsLinkLocalUnicast() && !addr.IsLinkLocalUnicast() {
+						inetAddr = addr
+					}
+					continue
+				}
+				inetAddr = addr
+			}
+			if enableInet6 {
+				if addr.Is6() {
+					// If we have already been set, can we improve on the quality?
+					if inet6Addr.IsValid() {
+						if inet6Addr.IsLinkLocalUnicast() && !addr.IsLinkLocalUnicast() {
+							inet6Addr = addr
+						}
+						continue
+					}
+					inet6Addr = addr
+				}
+			}
+		}
+	}
+
+	// If both address families have been set, compare.
+	if inetAddr.IsValid() && inet6Addr.IsValid() {
+		if inetAddr.IsLinkLocalUnicast() && !inet6Addr.IsLinkLocalUnicast() {
+			return inet6Addr
+		}
+		if inet6Addr.IsLinkLocalUnicast() && !inetAddr.IsLinkLocalUnicast() {
+			return inetAddr
+		}
+		if enableInet6 {
+			return inet6Addr
+		}
+		return inetAddr
+	}
+
+	if inetAddr.IsValid() {
+		return inetAddr
+	}
+
+	if enableInet6 {
```

Comment:
This check can be removed. If `!enableInet6` then `inet6Addr` should be never valid.